File: 프로그래머스/lv2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iterator>
#include <iostream>

using namespace std;

int solution(string str1, string str2) {
    int answer = 0;
    double similar = 0.0;
    vector<string> s1, s2;
    vector<string> uni, inter;
    
    for(int i = 0; i < str1.size(); i++)
        str1[i] = tolower(str1[i]);
    
    for(int i = 0; i < str2.size(); i++)
        str2[i] = tolower(str2[i]);
    
    for(int i = 0; i < str1.size() - 1; i++) 
        if(isalpha(str1[i]) && isalpha(str1[i + 1]))
           s1.push_back(str1.substr(i, 2));
    
    for(int i = 0; i < str2.size() - 1; i++)
        if(isalpha(str2[i]) && isalpha(str2[i + 1]))
           s2.push_back(str2.substr(i, 2));
    
    sort(s1.begin(), s1.end());
    sort(s2.begin(), s2.end());
    
    set_union(s1.begin(), s1.end(), s2.begin(), s2.end(), back_inserter(uni));
    set_intersection(s1.begin(), s1.end(), s2.begin(), s2.end(), back_inserter(inter));
    
    if(uni.empty())
        similar = 1 * 65536;
    else
        similar = (double)inter.size() / (double)uni.size() * 65536;
    
    return answer = similar;
}

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
from collections import Counter

def solution(str1, str2):
    _str1 = [str1[i:i+2].lower() for i in range(len(str1)-1) if str1[i:i+2].isalpha()]
    _str2 = [str2[i:i+2].lower() for i in range(len(str2)-1) if str2[i:i+2].isalpha()]
    
    if not _str1 and not _str2:
        return 65536
        
    c1, c2 = Counter(_str1), Counter(_str2)
    union = sum((c1|c2).values())
    inter = sum((c1&c2).values())
    
    return int(inter/union*65536)